protowhat/checks/check_bash_history.py

```diff
@@ -0,0 +1,181 @@
+    at the start of the exercise or every time the exercise is submitted.
+
+    Import using ``from protowhat.checks import update_bash_history_info``.
+    """
+    if bash_history_path is None:
+        bash_history_path = os.environ[BASH_HISTORY_PATH_ENV]
+    with open(bash_history_path, encoding="utf-8") as f:
+        Path(os.environ[BASH_HISTORY_INFO_PATH_ENV]).write_text(
+            str(len(f.readlines())), encoding="utf-8"
+        )
+
+
+def get_bash_history(
+    full_history=False, bash_history_path=None
+):
+    """Get the commands in the bash history
+
+    :param full_history: if true, returns all commands in the bash history,
+        else only return the commands executed after the last bash history info update
+    :param bash_history_path: path to the bash history file
+    :return: a list of commands (empty if the file is not found)
+
+    Import from ``from protowhat.checks import get_bash_history``.
+    """
+    if bash_history_path is None:
+        bash_history_path = os.environ[BASH_HISTORY_PATH_ENV]
+    try:
+        with open(bash_history_path, encoding="utf-8") as f:
+            history = f.readlines()
+            old_history_length = int(
+                Path(os.environ[BASH_HISTORY_INFO_PATH_ENV]).read_text()
+            )
+            return history if full_history else history[old_history_length:]
+    except FileNotFoundError:
+        return []
+
+
+"""
+Design considerations
+
+use case:
+bash history used multiple times in SCT
+solution:
+separate bash history info update
+
+use case:
+bash history is not empty from earlier exercise or session (that might not use this SCT)
+solution:
+update info in PEC or exercise init
+
+If updating bash history info in PEC, only use bash history in check_correct
+as it only has access to the latest commands
+"""
+
+
+def has_command(state, pattern, msg, fixed=False, commands=None):
+    """Test whether the bash history has a command matching the pattern
+
+    Args:
+        state: State instance describing student and solution code. Can be omitted if used with Ex().
+        pattern: text that command must contain (can be a regex pattern or a simple string)
+        msg: feedback message if no matching command is found
+        fixed: whether to match text exactly, rather than using regular expressions
+        commands: the bash history commands to check against,
+            by default all since the last bash history info update,
+            otherwise the result of calling the helper function
+            ``get_bash_history(full_history=False, bash_history_path=None)``
+            with other arguments can be passed.
+
+    Note:
+        The helper function ``update_bash_history_info(bash_history_path=None)``
+        needs to be called in the pre-exercise code in exercise types that don't have
+        support for bash history features by default.
+
+    Note:
+        If the bash history info is updated every time code is submitted,
```

Comment:
Suggestion, I realise it is verbose, but the usage is quite complex and different from other SCT's:
When ``update_bash_history_...`` is run in the pre-exercise-code, has_command should only be used in the second part of a check correct (to narrow down the error of the user), otherwise only the bash commands since the user's previous submission will be checked and they will have to rewrite all commands in an exercise with multiple commands, for example:
User needs to do ``ls`` to see subfolders and then ``cd`` into the solution folder:
```
User types:
>> ls
User submits, sct runs: 
  Ex().multi(
     has_command('ls'), 
     has_command('cd solution')
  )
--> Fails because cd solution was not found.
User types:
>> cd solution
User submits, sct runs: 
  Ex().multi(
     has_command('ls'), 
     has_command('cd solution')
  )
--> Fails because ls was not found! Using a check_correct here would let the user pass and still explains where the user went wrong:
Ex().check_correct(
  has_cwd('/home/repl/solution'),
  multi(
     has_command('ls'), 
     has_command('cd solution')
  )
)
```

Reply:
I added a second example.
In addition, if `update_bash_history_info` wasn't run, instructors will see an error.